packages/aws-cdk-lib/aws-s3-deployment/lib/render-data.ts

```diff
@@ -1,82 +1,44 @@
  */
-export function renderData(scope: Construct, data: string): Content {
-  const obj = Stack.of(scope).resolve(data);
-  if (typeof(obj) === 'string') {
-    return { text: obj, markers: {} };
-  }
+class TokenToMarkerMapper implements ITokenMapper {
+  markers: Record<string, any>;
 
-  if (typeof(obj) !== 'object') {
-    throw new ValidationError(`Unexpected: after resolve() data must either be a string or a CloudFormation intrinsic. Got: ${JSON.stringify(obj)}`, scope);
+  constructor() {
+    this.markers = {};
   }
 
-  let markerIndex = 0;
-  const markers: Record<string, FnJoinPart> = {};
-  const result = new Array<string>();
-  const fnJoin: FnJoin | undefined = obj['Fn::Join'];
-
-  if (fnJoin) {
-    const sep = fnJoin[0];
-    const parts = fnJoin[1];
-
-    if (sep !== '') {
-      throw new ValidationError(`Unexpected "Fn::Join", expecting separator to be an empty string but got "${sep}"`, scope);
-    }
-
-    for (const part of parts) {
-      if (typeof (part) === 'string') {
-        result.push(part);
-        continue;
-      }
-
-      if (typeof (part) === 'object') {
-        addMarker(part);
-        continue;
-      }
-
-      throw new ValidationError(`Unexpected "Fn::Join" part, expecting string or object but got ${typeof (part)}`, scope);
-    }
-  } else if (obj.Ref || obj['Fn::GetAtt'] || obj['Fn::Select']) {
-    addMarker(obj);
-  } else {
-    throw new ValidationError('Unexpected: Expecting `resolve()` to return "Fn::Join", "Ref" or "Fn::GetAtt"', scope);
+  mapToken(token: IResolvable) {
+    const newMarker = `<<marker:0xbaba:${Object.keys(this.markers).length}>>`;
+    this.markers[newMarker] = token.toString();
+    return newMarker;
   }
-
-  function addMarker(part: Ref | GetAtt | FnSelect) {
-    const keys = Object.keys(part);
-    const acceptedCfnFns = ['Ref', 'Fn::GetAtt', 'Fn::Select'];
-    if (keys.length !== 1 || !acceptedCfnFns.includes(keys[0])) {
-      const stringifiedAcceptedCfnFns = acceptedCfnFns.map((fn) => `"${fn}"`).join(' or ');
-      throw new ValidationError(`Invalid CloudFormation reference. Key must start with any of ${stringifiedAcceptedCfnFns}. Got ${JSON.stringify(part)}`, scope);
-    }
-
-    const marker = `<<marker:0xbaba:${markerIndex++}>>`;
-    result.push(marker);
-    markers[marker] = part;
-  }
-
-  return { text: result.join(''), markers };
 }
 
-type FnJoin = [string, FnJoinPart[]];
-type FnJoinPart = string | Ref | GetAtt | FnSelect;
-type Ref = { Ref: string };
-type GetAtt = { 'Fn::GetAtt': [string, string] };
-type FnSplit = { 'Fn::Split': [string, string | Ref] };
-type FnSelect = { 'Fn::Select': [number, string[] | FnSplit] };
+/**
+ * Renders the given string data as deployable content with markers substituted
+ * for all tokens.
+ *
+ * @param data The input data
+ * @returns The markered text (`text`) and a map that maps marker names to their
+ * values (`markers`).
+ */
+export function renderData(data: string): Content {
```

Comment:
Just noticed we don't have existing or new tests for this. IMO this would benefit from a unit test to prevent future regressions.

Reply:
Unit tests for `renderData` are already added in `content.test.ts`.

Reply:
Ah, my bad. Thanks for clarifying!